torch/_inductor/ordered_set.py

```diff
@@ -0,0 +1,129 @@
+from collections.abc import MutableSet, Set as AbstractSet
+from typing import Any, cast, Iterable, Iterator, Optional, TypeVar
+
+T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
+
+__all__ = ["OrderedSet"]
+
+
+class OrderedSet(MutableSet[T]):
+    """
+    Insertion ordered set, similar to OrderedDict.
+    """
+
+    def __init__(self, iterable: Optional[Iterable[T]] = None):
+        self._dict: dict[T, None] = {}
+        if iterable is not None:
+            self.update(iterable)
+
+    #
+    # Required overriden abstract methods
+    #
+
+    def __contains__(self, elem: Any) -> bool:
+        return elem in self._dict
+
+    def __iter__(self) -> Iterator[T]:
+        used = len(self)
+        it = iter(self._dict)
+        for _ in range(used):
+            if len(self) != used:
+                raise RuntimeError("OrderedSet changed size during iteration")
+            yield next(it)
+
+    def __len__(self) -> int:
+        return len(self._dict)
+
+    def add(self, elem: T) -> None:
+        self._dict[elem] = None
+
+    def discard(self, elem: T) -> None:
+        self._dict.pop(elem, None)
+
+    def clear(self) -> None:
+        # overridden because MutableSet impl is slwo
+        self._dict.clear()
+
+    # Unimplemented set() methods in _collections_abc.MutableSet
+
+    @classmethod
+    def _wrap_in_set(cls, other: Any) -> Any:
+        """
+        Wrap non-Set Iterables in OrderedSets
+
+        Some of the magic methods are more strict on input types than
+        the public apis, so we need to wrap inputs in sets.
+        """
+
+        if not isinstance(other, AbstractSet) and isinstance(other, Iterable):
+            return cls(other)
+        else:
+            return other
+
+    def pop(self) -> T:
+        if not self:
+            raise KeyError("pop from an empty set")
+        return self._dict.popitem()[0]
+
+    def copy(self) -> "OrderedSet[T]":
+        return self.__class__(self)
+
+    def difference(self, *others: Iterable[T]) -> "OrderedSet[T]":
+        res = self if len(others) else self.copy()
+        for other in others:
+            res = res - other  # type: ignore[operator, arg-type]
```

Comment:
This `type: ignore` seems suspect - I think the error is actually correct:
```
(base) $ python -c 'print(set((1, 2, 3)) - [2])'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: unsupported operand type(s) for -: 'set' and 'list'```
```
vs
```
(base) $ python -c 'print(set((1, 2, 3)).difference([2]))'
{1, 3}
```
So it seems that '-' actually doesn't support generic Iterables whereas difference does.  I guess they need _wrap_in_set as well.

(I assume the same thing applies to the following functions w/ `type: ignore` as well)

Reply:
I'm inheriting from collections.ABC.mutableset which allows this behavior. I mentioned this in the summary. See, this code here:

https://github.com/python/cpython/blob/3.12/Lib/_collections_abc.py#L642-L644

I would rather inherit from there behavior although we could decide to throw if want.

Reply:
This is fine - I didn't realize that MutableSet provided this functionality.  It's kind of weird that they support iterable but their signature takes AbstractSet.